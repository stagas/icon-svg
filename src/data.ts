import { Icons, IconTypes } from './types'

export const iconSets = {
  bootstrap: 'npm/bootstrap-icons@1/icons/{icon}.svg',
  boxicons: 'npm/boxicons@2/svg/{type}/{prefix}-{icon}.svg',
  bytesize: 'npm/bytesize-icons@1/dist/icons/{icon}.svg',
  cssgg: 'npm/css.gg@2/icons/svg/{icon}.svg',
  emojicc: 'npm/emoji-cc@1/svg/{icon}.svg',
  eos: 'gh/lekoala/eos-icons-mirror/{type}/{icon}.svg',
  feather: 'npm/feather-icons@4/dist/icons/{icon}.svg',
  flags: 'npm/flag-svg-collection@1/flags/{kind}/{icon}.svg',
  fontawesome: 'npm/@fortawesome/fontawesome-free@5/svgs/{type}/{icon}.svg',
  iconoir: 'gh/lucaburgio/iconoir/icons/{icon}.svg',
  iconpark: 'gh/bytedance/IconPark/source/{type}/{icon}.svg',
  material: 'npm/@material-icons/svg@1/svg/{icon}/{kind}.svg',
  supertiny: 'npm/super-tiny-icons@0.4.0/images/svg/{icon}.svg',
  tabler: 'npm/@tabler/icons@1/icons/{icon}.svg',
}

export const prefixes = {
  boxicons: {
    solid: 'bxs',
    regular: 'bx',
    logos: 'bxl',
  },
}

export type IconSetsNames = keyof typeof iconSets
export type IconSets<T extends IconSetsNames> = { [k in T]: string }
export type IconsWithTypes = Pick<Icons, keyof IconTypes>
