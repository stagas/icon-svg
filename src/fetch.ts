import { memoize } from 'memoize-pure'
import { getIconUrl, IconSets, IconSetsNames, IconsWithTypes, Icons, IconTypes, IconKinds } from '.'

const fetchMemoized = memoize(async (url: string) => {
  const response = await fetch(url)
  const text = await response.text()
  if (response.ok) return text
  else {
    console.error(url + ': ' + text)
    return `<span title="Error loading icon: ${url}">⚠️</span>`
  }
})

export interface FetchProps {
  raw?: boolean
  strokeWidth?: string | number
}

/**
 * Fetches an SVG icon.
 *
 * The network request is memoized so when this is called multiple times
 * it will only access the network the first time (or the browser cache if there already).
 *
 * It also attempts to normalize the SVG to make all the icons consistent and
 * usable out of the box without any adjustments, i.e it normalizes various values.
 *
 * Use CSS variables `--stroke` `--fill` and `--fill-secondary` to manage
 * the colors if necessary.
 *
 * When used with `strokeWidth` you can achieve fine-control on how it's being displayed.
 *
 * Use the `raw` option to skip any post processing if you don't need it and simply get
 * the raw SVG data.
 *
 * ```ts
 * svg = await fetchIconSvg({ set: 'bytesize', icon: 'heart', strokeWidth: '2px' })
 * svg = await fetchIconSvg({ set: 'flags', icon: 'gr', kind: '4x3', raw: true })
 * svg = await fetchIconSvg({
 *   set: 'boxicons',
 *   icon: 'alarm',
 *   type: 'solid',
 *   strokeWidth: '' // passing `strokeWidth=''` allows it to be managed in CSS
 * })
 * ```
 *
 * @param props
 * @param props.set Icon set
 * @param props.icon Icon name
 * @param props.type Icon type - Most of the time an entirely different set of icons
 * @param props.kind Icon kind - These are variations of the same icon
 * @param props.raw Get the raw SVG without any post processing
 * @param props.strokeWidth Normalize `stroke-width` to this value.
 *  This also applies `vector-effect="non-scaling-stroke"` which will make the
 *  `stroke-width` absolute and not relative to the size of the icon
 *  Use this option for more control over the icon's appearance.
 *  Because every dimension of the icon will need a different `stroke-width`
 *  this can be set to `''` (empty string) which then allows it to be handled with CSS.
 *
 * @returns The SVG data as a string
 */
export const fetchIconSvg = async <P extends keyof IconSets<IconSetsNames>, T extends IconTypes[P & keyof IconTypes]>(
  props: FetchProps & {
    set: P
  } & (P extends keyof IconKinds
      ? {
          icon: Icons[P]
          kind: IconKinds[P & keyof IconKinds]
        }
      : P extends keyof IconTypes
      ? {
          icon: IconsWithTypes[P & keyof IconsWithTypes][T & keyof IconsWithTypes[P & keyof IconsWithTypes]]
          type: T
        }
      : {
          icon: Icons[P & keyof Icons]
        })
) => {
  const url = getIconUrl(props as never)
  let svg = (await fetchMemoized(url)).trim()

  if (props.raw || ['emojicc', 'flags'].includes(props.set)) return svg

  svg = svg
    .replace(/(?<=^<svg[^>]*)\s(width|height)="[\d.]+"(?=.*>)/gis, '')
    .replace(/(^<svg[^>]*?)>/gis, '$1 fill="var(--stroke)">')

    // these two are weird unnecessary squares in iconpark's icons
    .replace('<rect width="48" height="48" fill="white" fill-opacity="0.01"/>', '')
    .replace('<path d="M48 0H0V48H48V0Z" fill="white" fill-opacity="0.01"/>', '')

  // normalize stroke-width
  if (props.strokeWidth != null)
    svg = svg
      .replace(/(stroke-width=")([^"]*?)(")/gi, `$1${props.strokeWidth as string}$3`)
      .replace(/(<(?:path|circle|rect|polyline)[^>]*?)(\/?>)/gis, '$1 vector-effect="non-scaling-stroke"$2')

  svg = svg
    // these are iconpark's colors but could be extended further
    .replace(/"#2F88FF"/gis, '"var(--fill)"')
    .replace(/"#43CCF8"/gis, '"var(--fill-secondary)"')

    // inherit so we can style with css
    .replace(/"(currentColor|white|black)"/gis, '"var(--stroke)"')

  return svg
}
