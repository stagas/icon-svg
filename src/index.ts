export * from './component'
export * from './data'
export * from './element'
export * from './fetch'
export * from './icon-svg'
export * from './types'
