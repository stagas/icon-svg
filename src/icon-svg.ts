import { IconSets, iconSets, IconSetsNames, IconsWithTypes, prefixes } from './data'
import { IconKinds, Icons, IconTypes } from './types'

const origin = 'https://cdn.jsdelivr.net'

/**
 * Gets the SVG CDN url of an icon.
 *
 * ```ts
 * getIconUrl({ set: 'bytesize', icon: 'heart' })
 * getIconUrl({ set: 'boxicons', icon: 'alarm', type: 'solid' })
 * getIconUrl({ set: 'flags', icon: 'gr', kind: '4x3' })
 * ```
 *
 * @param props
 * @param props.set Icon set
 * @param props.icon Icon name
 * @param props.type Icon type - Most of the time an entirely different set of icons
 * @param props.kind Icon kind - These are variations of the same icon
 * @returns The SVG CDN url
 */
export const getIconUrl = <P extends keyof IconSets<IconSetsNames>, T extends IconTypes[P & keyof IconTypes]>(
  props:
    & {
      set: P
    }
    & (P extends keyof IconKinds ? {
        icon: Icons[P]
        kind: IconKinds[P & keyof IconKinds]
      }
      : P extends keyof IconTypes ? {
          icon: IconsWithTypes[P & keyof IconsWithTypes][T & keyof IconsWithTypes[P & keyof IconsWithTypes]]
          type: T
        }
      : {
        icon: Icons[P & keyof Icons]
      }),
) => {
  if (!propsSatisfyTemplate(props)) throw new TypeError('IconSvg: Missing properties')
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const prefix = prefixes[props.set as never]?.[(props as any).type]
  const values = Object.assign({}, props, { prefix })
  let url = origin + '/' + iconSets[props.set].replace(/{(.*?)}/g, (_, key) => values[key as never] as string)
  if ((props as any).set === 'phosphor' && (props as any).type === 'regular') url = url.replace('-regular', '')
  return url
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const propsSatisfyTemplate = (props: any) => {
  if (!('set' in props) || props.set == null) return false
  const template = iconSets[props.set as never] as string
  const keys = [...template.matchAll(/{(.*?)}/g)].map(x => x[1])
  for (const key of keys) {
    if (key === 'prefix') continue
    if (!(key in props) || props[key] == null) return false
  }
  return true
}
