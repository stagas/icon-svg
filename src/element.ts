import { attrs, mixter, shadow } from 'mixter'
import { fetchIconSvg } from './fetch'
import { propsSatisfyTemplate } from './icon-svg'

const style = ({ strokeWidth }: { strokeWidth?: string | number | null }) => /*css*/ `
:host {
  ${strokeWidth ? `stroke-width: ${strokeWidth};` : ''}
  --stroke: currentColor;
  --fill: #9994;
  --fill-secondary: #9996;
  display: inline-flex;
  vertical-align: middle;
}

[part=svg] {
  width: 100%;
  height: 100%;
}`

/**
 * The `IconSvgElement` custom element.
 *
 * For a more complete example with typings check `playground/app.tsx`
 *
 * Browser needs to support `Constructible StyleSheets` ([polyfill](https://github.com/calebdwilliams/construct-style-sheets)).
 *
 * _JS / TS:_
 * ```ts
 * customElements.define('icon-svg', IconSvgElement)
 * ```
 *
 * _CSS:_
 * ```css
 * icon-svg {
 *   width: 32px;
 *   height: 32px;
 *   stroke-width: 2.5px;
 *   --stroke: #fff;
 *   --fill: #028;
 *   --fill-secondary: #468;
 * }
 * ```
 *
 * _HTML:_
 * ```html
 * <icon-svg set="boxicons" type="logos" icon="javascript"></icon-svg>
 * ```
 */
export class IconSvgElement extends mixter(
  HTMLElement,
  shadow(),
  attrs(
    class {
      icon = String
      set = String
      type = String
      kind = String
      raw = Boolean
      strokeWidth = Number
    }
  )
) {
  async updateSvg() {
    const props = {
      set: this.getAttribute('set'),
      icon: this.getAttribute('icon'),
      type: this.getAttribute('type'),
      kind: this.getAttribute('kind'),
      raw: this.hasAttribute('raw'),
      strokeWidth: this.getAttribute('stroke-width'),
    }
    if (!propsSatisfyTemplate(props)) return

    this.shadowRoot!.innerHTML = `<style>${style(props)}</style>${
      (await fetchIconSvg(props as never)).replace(/(<svg.*?)>/, '$1 part="svg">')
    }`
  }

  attributeChangedCallback() {
    this.updateSvg()
  }
}
