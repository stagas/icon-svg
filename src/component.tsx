/** @jsxImportSource html-vdom */
import { fromElement } from 'html-vdom/from-element'
import { IconKinds, Icons, IconSets, IconSetsNames, IconsWithTypes, IconTypes } from './'
import { IconSvgElement } from './element'

const IconSvgTag = fromElement(IconSvgElement)

export const IconSvg = <P extends keyof IconSets<IconSetsNames>, T extends IconTypes[P & keyof IconTypes]>(
  props: P extends keyof IconKinds ? {
    set: P
    icon: Icons[P]
    kind: IconKinds[P & keyof IconKinds]
  }
    : P extends keyof IconTypes ? {
      set: P
      icon: IconsWithTypes[P & keyof IconsWithTypes][T & keyof IconsWithTypes[P & keyof IconsWithTypes]]
      type: T
    }
    : {
      set: P
      icon: Icons[P & keyof Icons]
    },
) => <IconSvgTag {...props} stroke-width></IconSvgTag>

IconSvg.toString = IconSvgTag.toString
