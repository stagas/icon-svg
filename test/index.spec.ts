import { inspectWithPreamble, setOptions } from '@n1kk/intspector'
import { getIconUrl } from '../src/icon-svg'

setOptions(require('../tsconfig.json'), true)

const typeTest = (fn: string) => {
  try {
    inspectWithPreamble(`
      import { getIconUrl } from '../'
      const result = ${fn}
    `)({ result: 'typeof result' })
    return true
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    return error[0].messageText.next[0].messageText
  }
}

describe('getIconUrl', () => {
  it('only package + name', () => {
    expect(getIconUrl({ set: 'feather', icon: 'award' })).toMatchSnapshot()
    expect(getIconUrl({ set: 'bytesize', icon: 'heart' })).toMatchSnapshot()
  })

  it('package + name + type', () => {
    expect(getIconUrl({ set: 'flags', icon: 'gr', kind: '4x3' })).toMatchSnapshot()
  })

  it('package + name + type adjusts prefix', () => {
    expect(getIconUrl({ set: 'boxicons', icon: 'alarm', type: 'solid' })).toMatchSnapshot()
  })

  it('throws when missing properties', () => {
    expect(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      getIconUrl({})
    }).toThrow('Missing properties')
  })

  it('missing type when required', () => {
    expect(typeTest(`getIconUrl({ set: 'boxicons', type: 'logos', icon: '99designs' })`)).toEqual(true)
    expect(typeTest(`getIconUrl({ set: 'boxicons', icon: '99designs' })`)).toContain('\'type\' is missing')
  })

  it('missing kind when required', () => {
    expect(typeTest(`getIconUrl({ set: 'flags', icon: 'gr', kind: '4x3' })`)).toEqual(true)
    expect(typeTest(`getIconUrl({ set: 'flags', icon: 'gr' })`)).toContain('\'kind\' is missing')
  })
})
