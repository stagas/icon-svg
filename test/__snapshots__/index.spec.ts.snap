// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIconUrl only package + name 1`] = `"https://cdn.jsdelivr.net/npm/feather-icons@4/dist/icons/award.svg"`;

exports[`getIconUrl only package + name 2`] = `"https://cdn.jsdelivr.net/npm/bytesize-icons@1/dist/icons/heart.svg"`;

exports[`getIconUrl package + name + type 1`] = `"https://cdn.jsdelivr.net/npm/flag-svg-collection@1/flags/4x3/gr.svg"`;

exports[`getIconUrl package + name + type adjusts prefix 1`] = `"https://cdn.jsdelivr.net/npm/boxicons@2/svg/solid/bxs-alarm.svg"`;
